Always 5 rocks; Remove Black Omen Spot; fix_flag_conflicts unit testing

coffeemancy opened the pull request:
This PR updates Rocksanity to always add 5 rock KIs regardless of game mode and provide a flag (`REMOVE_BLACK_OMEN_ROCK`) to remove the Black Omen Terra Rock chest as a possible location for KIs. This is useful for non-troll Rock Hunt (Bucket List with "Collect 5 Rocks" objective).

It also involves changes around `fix_flag_conflicts`, which counts KIs/spot, and automatically adjusts for differences by adding Vanilla Robo Ribbon or removing Epoch Fail.

## Updates

### User-Facing

* Always adds all 5 rocks as KIs when Rocksanity is used, regardless of game mode
* Adds new flag `REMOVE_BLACK_OMEN_SPOT` which removes Black Omen Terra Rock chest from pool of KI locations (which only really impacts Standard play with Rocksanity at the moment; forced off for LW)
* Forces off `ADD_BEKKLER_SPOT` for Ice Age, Legacy of Cyrus (no CTrigger); `ADD_RACELOG_SPOT` for LoC (no Future)
* Updates counting/adjustment of flags in `fix_flag_conflicts` to account for Chronosanity (don't need to adjust flags as plenty of spots), Rocksanity, Ice Age and Legacy of Cyrus (no black omen, more extra spots to fill with KIs), and new REMOVE_BLACK_OMEN_SPOT flag
* Tracks initial flag settings and displays a diff of flags in spoiler log if flags were added or removed (i.e. by `fix_flag_conflicts`); split and wrap flags at 80 characters per line
* TK GUI gracefully handles errors during rom generation, showing error in popup (instead of crashing)

### Development Updates

* Adds `flake8` configuration file, uses it; constrains test requirements
* Adds unit testing of `fix_flag_conflicts`, run on PRs via `pytest`

## Testing

* [Github Actions](https://github.com/coffeemancy/jetsoftime/actions/runs/6113898462)

### Web GUI

Using https://github.com/Anguirel86/ctjot_web_generator/pull/20

Standard with Epoch Fail, Rocksanity, Remove Black Omen Spot, Restore Johnny Race. That's +7 items, only +4 spots, so `fix_flag_conflicts` will add Robo Ribbon and remove Epoch Fail.

Spoiler shows flags include Remove Black Omen Spot and diff shows Vanilla Robo Ribbon added and Epoch Fail removed, no "Black Omen Terra Rock" spot for Key Items:

[no-black-omen-web-gui.webm](https://github.com/coffeemancy/jetsoftime/assets/3493534/cae5902f-1da0-437e-a332-857c52fab1b9)

Generate seed without Remove Black Omen Spot shows spot in spoiler still:

![no-black-omen-yes-black-omen](https://github.com/coffeemancy/jetsoftime/assets/3493534/9be6c673-a9f2-46ee-af99-84f651eb51aa)

### CLI

Legacy of Cyrus with Epoch Fail, Rocksanity, Duplicate Chars. That's +6 KI / +4 spots, but LoC allows for extra one (fewer base KI), so no changes, as shown in spoiler:

```bash
python randomizer.py -i ct.sfc --json-spoilers --spoilers --mode=loc -ef --rocksanity -dc
```

Spoiler shows diff in flags adds Unlocked Skygates (for Rocksanity) and Char Rando (for Duplicate Chars), Epoch Fail remains, no Vanilla Robo Ribbon added:

```
Flags: GameFlags.ROCKSANITY|CHAR_RANDO|DUPLICATE_CHARS|EPOCH_FAIL|UNLOCKED_SKYGATES
       Diff:
         + GameFlags.CHAR_RANDO|UNLOCKED_SKYGATES
```

JSON spoiler shows initial flags as well (`jq '.settings'`):

```json
{
  "seed": "metalmuterolypoly",
  "mode": "Legacy of cyrus",
  "enemy_difficulty": "Normal",
  "item_difficulty": "Normal",
  "tech_order": "Full random",
  "shops": "Normal",
  "flags": [
    "GameFlags.ROCKSANITY",
    "GameFlags.CHAR_RANDO",
    "GameFlags.DUPLICATE_CHARS",
    "GameFlags.EPOCH_FAIL",
    "GameFlags.UNLOCKED_SKYGATES"
  ],
  "initial_flags": [
    "GameFlags.ROCKSANITY",
    "GameFlags.DUPLICATE_CHARS",
    "GameFlags.EPOCH_FAIL"
  ],
  "cosmetic_flags": []
}
```

### TK GUI

1. Standard with Epoch Fail, Rocksanity, Unlocked Skygates, Restore Johnny Race, Restore Tools, Remove Black Omen Spot (+8 KIs, +4 spots).
2. Click "Generate". GUI shows error message and doesn't crash.
3. Remove Epoch Fail. Click "Generate". Still fails (`fix_flag_conflicts` can't remove it, as already removed).
4. Add Cyrus Spot. Click "Generate". Works.

[no-black-omen-tk-gui.webm](https://github.com/coffeemancy/jetsoftime/assets/3493534/ec75123b-e009-401a-a8ef-1b5906bf4814)

Spoiler shows flags include Remove Black Omen Spot and Vanilla Robo Ribbon (shown also in diff), and the spot has an item:

```
Flags: GameFlags.FIX_GLITCH|ZEAL_END|FAST_PENDANT|ROCKSANITY|UNLOCKED_SKYGATES
                |ADD_CYRUS_SPOT|RESTORE_TOOLS|RESTORE_JOHNNY_RACE
                |VANILLA_ROBO_RIBBON|REMOVE_BLACK_OMEN_SPOT
       Diff:
         + GameFlags.VANILLA_ROBO_RIBBON
...
Black Omen Terra Rock                         Crisis Arm
```

Pseudoarc commented:
The idea of having all five rocks in other game modes is good.

I'm not sure how I feel about making the vanilla white rock placement impossible for the sake of the bucket list objective.  Some people like the outside chance of troll seeds, but I agree that it'd be nice to have a way to disable that behavior.   Removing the spot would also mean that Rocksanity would need to be added to the list of flags that remove a KI spot/add a KI (+5 KIs, +4 spots) and be included in the ki/spot validation.  As a result EF+Rocksanity without any extra spots would be invalid (unless I'm missing something), and I don't think that's desirable.

It seems natural to have a sub-flag for Rocksanity to control this behavior (like duplicate techs with random characters).  What do you think?

coffeemancy commented:
Discussed on discord.

Going to update PR:

* Keep Black Omen Terra Rock location for Rocksanity by default (in Standard)
* Add new flag `REMOVE_BLACK_OMEN_SPOT` to remove that spot which can be separately enabled
* Update CLI/TK GUI to note about options that "add/remove a spot" (currently we only have spots that "add")

Commit:
always 5 rocks

Don't remove a rock from KI pool anymore with Legacy of Cyrus,
Ice Age, etc., as they already have extra spots, and rocks can
replace treasure/Jerky fills.

Commit:
move tests to tests directory

Use same hack of appending to import path with sourcefiles directory
as is used in ctjot_web_generator for now. In future, could be nice
to convert to python package format and use native python importing
here and in other repos.

Commit:
flake8 config based on existing code, constrain test requirements

Went through code and organized all current flake8 failures and
added to ignore in .flake8. Some are issues that probably should
be addressed soon, others are minor formatting, and some contributors
just might not care about (trying not to be dogmatic). Went with 119
line-length because almost all code in repo has less than that (and
quite a bit has more than 79 characters, which is another common
boundary for python projects).

Commit:
very low-risk fixes for minor one-off issues flake8 caught

Commit:
add unit test coverage for fix_flag_conflicts, handle rocksanity

Adds some unit tests for fix_flag_conflicts logic for different
modes, KI additions/rocksanity, and spots.

Commit:
fix LoC/IA forced_off_dict, add spots based on mode, chronosanity

Fix LoC to disable flags requiring CTrigger or future access,
IA to disable flags requiring CTrigger (Bekkler). Add 2 extra
effective "spots" for IA (1 for LoC) in fix_flag_conflicts
because some treasure spots can be filled with KIs. LoC only
gets 1 incase Marle/Robo at Proto. Do not adjust flags based on
KIs for Chronosanity because there are plenty of spots.

Commit:
align and split spoiler printing flags, print flag diff

Align text output for spoiler log with flags spanning multiple
lines, instead of one super long line. Print a diff from initial
flags versus used in generation, if different (e.g. when
fix_flag_conflicts adjusts them).

Minor flake8 cleanups.

Commit:
add REMOVE_BLACK_OMEN_SPOT flag and test it

Adds new flag REMOVE_BLACK_OMEN_SPOT. When used, removes Black
Omen Terra Rock spot from KI pool. Currently only affects Rocksanity.
Update tests to cover it.

Commit:
gracefully handle exceptions and show error in TK GUI

And extend to 22 width for new flag text to not be truncated.

Commit:
add python 3.8 back in testing matrix, actively used
